onnx/test/runtime_test.py

```diff
@@ -0,0 +1,1174 @@
+                [15, 59, 237, 95, 129, 0, 64],
+                [56, 242, 153, 221, 168, 12, 166],
+                [232, 178, 186, 195, 237, 162, 237],
+                [188, 39, 124, 77, 80, 102, 43],
+                [127, 230, 21, 83, 41, 40, 134],
+                [255, 154, 92, 141, 42, 148, 247],
+            ],
+            dtype=np.uint8,
+        ).reshape((1, 1, 7, 7))
+        x_scale = np.array([0.00369204697], dtype=np.float32)
+        x_zero_point = np.array([132], dtype=np.uint8)
+        w = np.array([0], dtype=np.uint8).reshape((1, 1, 1, 1))
+        w_scale = np.array([0.00172794575], dtype=np.float32)
+        w_zero_point = np.array([255], dtype=np.uint8)
+        y_scale = np.array([0.00162681262], dtype=np.float32)
+        y_zero_point = np.array([123], dtype=np.uint8)
+
+        feeds = {
+            "x": x,
+            "x_scale": x_scale,
+            "x_zero_point": x_zero_point,
+            "w": w,
+            "w_scale": w_scale,
+            "w_zero_point": w_zero_point,
+            "y_scale": y_scale,
+            "y_zero_point": y_zero_point,
+        }
+        expected = sess1.run(None, feeds)[0]
+        got = sess2.run(None, feeds)[0]
+        assert_almost_equal(expected, got)
+
+    def common_test_im2col(self, kernel_shape, pads, strides, dilations):
+        X = make_tensor_value_info("X", TensorProto.FLOAT, [None, None, None, None])
+        Y1 = make_tensor_value_info("Y1", TensorProto.FLOAT, [None, None, None, None])
+        Y2 = make_tensor_value_info("Y2", TensorProto.FLOAT, [None, None, None, None])
+        W = make_tensor_value_info("W", TensorProto.FLOAT, [None, None, None, None])
+        node = make_node(
+            "Conv", ["X", "W"], ["Y1"], pads=pads, strides=strides, dilations=dilations
+        )
+        node_shape = make_node("Shape", ["W"], ["shape"])
+        node_im = make_node(
+            "Im2Col",
+            ["X", "shape"],
+            ["xim"],
+            pads=pads,
+            strides=strides,
+            dilations=dilations,
+            domain="experimental",
+        )
+        node_flat = make_node("Flatten", ["W"], ["wflat"])
+        node_axes = make_node(
+            "Constant", [], ["axes"], value=from_array(np.array([0], dtype=np.int64))
+        )
+        node_qu = make_node("Squeeze", ["wflat", "axes"], ["wsqu"])
+        node_gem = make_node("MatMul", ["xim", "wsqu"], ["Y2"])
+        graph = make_graph(
+            [node, node_shape, node_im, node_axes, node_flat, node_qu, node_gem],
+            "g",
+            [X, W],
+            [Y1, Y2],
+        )
+        onnx_model = make_model(
+            graph, opset_imports=[make_opsetid("", 16), make_opsetid("experimental", 1)]
+        )
+        graph_conv = make_graph([node], "g", [X, W], [Y1])
+        onnx_model_conv = make_model(graph_conv, opset_imports=[make_opsetid("", 16)])
+        sess = rt.Inference(onnx_model)
+
+        try:
+            import onnxruntime as ort
```

Comment:
I am a bit unclear why we have tests of ort here?

Reply:
They can be removed. I'm just checking that the runtime I write and onnxruntime have the same results. Backend tests are not enough to understand what an implementation does wrong.